oeps/oep-0032-arch-unique-identifier-for-users.rst

```diff
@@ -0,0 +1,127 @@
+------------------------------------
+
+The following table summarizes the comparisons of a list of candidate user identifiers.
+
+.. list-table::
+   :header-rows: 1
+   :widths: 25 75
+
+   * - Identifier
+     - Comments
+   * - username
+     - * Username is considered PII.
+   * - LMS user_id
+     - * anonymous (is not considered PII)
+
+       * Segment (used for anlytics) recommends database IDs because they never change.
+
+       * Availability across all services is almost, but not yet complete.
+
+       * Auto-incremented id allows for possible information leakage of auth_user table and ties us down to implementation of the housing database.
+   * - email address
+     - * Is PII and modifiable by the user.
+   * - anonymous user_id
+     - * anonymous (is not considered PII)
+
+       * It is currently constructed by hashing the user's *LMS user_id* with the Django server's *SECRET_KEY* value. This value would change when the *SECRET_KEY* is rotated.
+
+As noted above, the *LMS user_id* was selected for internal analytics events as the best match given the requirements. It continues to be the best match for the additional use cases of public events, REST APIs and JavaScript APIs.
+
+
+Consequences
+============
+
+A consequence of the decision to use the *LMS user_id* as the unique user id is that some additional changes are needed to make the *LMS user_id* more globally available across all services:
+
+* The `LMS OAuth Dispatch App`_ will now include the *LMS user_id* in the JWTs (JSON Web Tokens) used as part of the `Open edX Authentication`_ upon request.
+
+* In order to keep the *LMS user_id* anonymous, there must not be any APIs or pages that _publicly_ expose the association between the *LMS user_id* and any of the user's PII (e.g. username and full name).
+
+* Additionally, some further investment is required to make the *LMS user_id* available to all services for all users, like the Ecommerce Service.
+
+This decision is also a simple acknowledgement of our tie to the database implementation and its auto-incremented value, given its already pervasive usage.
+
+The *LMS user_id* is also considered a safe and anonymous option for third-party integrations as described above. One example of a third-party integration where the *LMS user_id* can now safely be used is in the `Realtime Events API`_, which has been updated with this clarification.
+
+Note: The legacy implementation of LTI uses what is called an "anonymous id", but is no more "anonymous" than the *LMS user_id* from a PII perspective.  It uses an id made of a combination of user id and course id, which makes it impossible for a third party system to build a user model in its own system across courses, and limits the capabilities a third-party system might offer a user. For future LTI implementations, this decision simply states that the *LMS user_id* is an option unless these additional limitations are a requirement of the integration.
```

Comment:
```suggestion
Note: The legacy implementations of LTI and the %%USER_ID%% keyword uses what is called an "anonymous id", but is no more "anonymous" than the *LMS user_id* from a PII perspective. They use an id made of a combination of user id and course id, which makes it impossible for a third party system to build a user model in its own system across courses, and limits the capabilities a third-party system might offer a user. For future LTI implementations, this decision simply states that the *LMS user_id* is an option unless these additional limitations are a requirement of the integration.
```

Reply:
@robrap I agree with @pdpinch that the `%%USER_ID%%` is an additional example of a legacy feature that can be changed with future implementations, given the decision in this OEP.

One suggestion: to minimize the possible confusion, you can refer to `%%USER_ID%%` as an additional example (in a separate note) after this LTI note.